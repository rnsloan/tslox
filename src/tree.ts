// deno-lint-ignore no-explicit-any
type Data = any;

enum STRATEGY {
  DEPTH,
  BREADTH,
}

export interface INode {
  parent: Node | null;
  children: Node[];
  data: Data;
  isRoot: () => boolean;
  isLeaf: () => boolean;
  hasChildren: () => boolean;
  drop: () => void;
  addChild: (data: Data) => Node;
  addChildAtIndex: (data: Data, index: number) => Node;
  getPath: (node: Node) => Node[];
}

class Node implements INode {
  protected id: string;
  readonly parent: Node | null;
  readonly children: Node[];
  readonly data: Data;

  constructor(data: Data, parent?: Node) {
    this.id = crypto.randomUUID();
    this.parent = parent || null;
    this.data = data;
    this.children = [];
  }

  isRoot() {
    return false;
  }

  isLeaf() {
    return Boolean(!this.children.length);
  }

  hasChildren() {
    return Boolean(this.children.length);
  }

  drop(): Node | undefined {
    if (this.isRoot()) {
      console.warn("Cannot drop the Root Node");
      return;
    }

    const index = this.parent?.children.findIndex((n) => n.id === this.id);

    if (index) {
      const result = this.parent?.children.splice(index, 1);
      if (result?.length) {
        return result[0];
      }
    }

    console.error("Unable to drop Node");
  }

  addChild(data: Data) {
    const node = new Node(data, this);
    this.children.push(node);
    return node;
  }

  addChildAtIndex(data: Data, index: number) {
    const node = new Node(data, this);
    this.children.splice(index, 0, node);
    return node;
  }

  getPath() {
    interface AddParent {
      (n: Node): AddParent;
    }

    const path = [];
    path.push(this);

    const addParent = (n: Node): AddParent | undefined => {
      const { parent } = n;

      if (parent) {
        path.push(parent);
        return addParent(parent);
      }
    };

    addParent(this);
    return path.reverse();
  }
}

export interface IRootNode extends INode {
  find: (
    predicate: (n: Node) => boolean,
    strategy?: STRATEGY,
  ) => Node | undefined;
  findAll: (predicate: (n: Node) => boolean, strategy?: STRATEGY) => Node[];
}

export class RootNode extends Node implements IRootNode {
  constructor(data: Data) {
    super(data);
    this.id = `root-${crypto.randomUUID()}`;
  }

  isRoot() {
    return true;
  }

  isLeaf() {
    return false;
  }

  find(predicate: (n: Node) => boolean, strategy?: STRATEGY): Node | undefined {
    let result: Node | undefined;

    const findDfs = (node: Node) => {
      if (predicate(node) && !result) {
        result = node;
        return;
      }

      for (let i = 0; i < node.children.length; i++) {
        findDfs(node.children[i]);
      }
    };
    const findBfs = (node: Node) => {
      const queue = [node];

      while (queue.length && !result) {
        const n = queue.shift();

        if (n && predicate(n)) {
          result = n;
        }

        n?.children?.forEach((item) => queue.push(item));
      }
    };

    if (strategy === STRATEGY.BREADTH) {
      findBfs(this);
    } else {
      findDfs(this);
    }

    return result;
  }

  findAll(predicate: (n: Node) => boolean, strategy?: STRATEGY): Node[] {
    const result: Node[] = [];
    const findAllDfs = (node: Node) => {
      if (predicate(node)) {
        result.push(node);
      }

      node.children.forEach((n) => {
        findAllDfs(n);
      });
    };

    const findAllBfs = (node: Node) => {
      const queue = [node];

      while (queue.length) {
        const n = queue.shift();

        if (n && predicate(n)) {
          result.push(n);
        }

        n?.children?.forEach((item) => queue.push(item));
      }
    };

    if (strategy === STRATEGY.BREADTH) {
      findAllBfs(this);
    } else {
      findAllDfs(this);
    }

    return result;
  }
}

export default class Tree {
  static STRATEGY = STRATEGY;

  root: RootNode | null;

  constructor() {
    this.root = null;
  }

  parse(data: Data) {
    this.root = new RootNode(data);
    return this.root;
  }
}
