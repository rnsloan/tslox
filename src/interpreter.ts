import { Tree } from "./tree.ts";
import { ASTNode, ASTNodeType, IProgram } from "./parser.ts";

function convertTreeToASTTree(tree: Tree): IProgram {
  const astTree: IProgram = {
    type: ASTNodeType.Program,
    body: [],
  };

  tree.root?.children.map((child) => astTree.body.push(child.data));

  return astTree;
}

function generateCode(node: ASTNode): string {
  let code = "";

  switch (node.type) {
    case ASTNodeType.Program: {
      code = node.body.map((n) => generateCode(n)).join("\n");
      break;
    }
    case ASTNodeType.VariableDeclaration: {
      code = `var ${node.declarations[0].id.name} = ${
        generateCode(node.declarations[0].init)
      };`;
      break;
    }
    case ASTNodeType.Identifier: {
      code = node.name;
      break;
    }
    case ASTNodeType.Literal: {
      if (node.value) {
        code = `${node.value}`;
      } else if (node.raw === null) {
        code = "null";
      } else {
        code = `"${node.raw}"`;
      }
      break;
    }
    case ASTNodeType.CallExpression: {
      let args = "";
      if (node.arguments) {
        args = node.arguments.map((arg) => generateCode(arg)).join(",");
      }
      code = `${node.callee.name}(${args})`;
      break;
    }
    default: {
      throw new Error(`Unrecognised AST type ${node.type}`);
    }
  }

  return code;
}

export function interpreter(ast: Tree): string {
  const astTree = convertTreeToASTTree(ast);
  return generateCode(astTree);
}
