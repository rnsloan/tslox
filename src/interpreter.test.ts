import { assertSnapshot } from "https://deno.land/std@0.143.0/testing/snapshot.ts";
import { Tree } from "./tree.ts";
import { interpreter } from "./interpreter.ts";

Deno.test("Variables", async (t) => {
  const tree = new Tree();
  const root = tree.parse({ type: "Program" });
  root.addChildren([{
    type: "VariableDeclaration",
    declarations: [
      {
        type: "VariableDeclarator",
        id: { type: "Identifier", name: "hello" },
        init: { type: "Literal", value: 12.45, raw: "12.45" },
      },
    ],
  }, {
    type: "VariableDeclaration",
    declarations: [
      {
        type: "VariableDeclarator",
        id: { type: "Identifier", name: "world" },
        init: { type: "Literal", value: null, raw: null },
      },
    ],
  }]);

  const result = interpreter(tree);
  await assertSnapshot(t, result);
});

Deno.test("Functions", async (t) => {
  const tree = new Tree();
  const root = tree.parse({ type: "Program" });
  root.addChildren([
    {
      type: "CallExpression",
      callee: { type: "Identifier", name: "foo" },
      arguments: [{ type: "Literal", value: true, raw: "true" }],
    },
    {
      type: "VariableDeclaration",
      declarations: [
        {
          type: "VariableDeclarator",
          id: { type: "Identifier", name: "foo" },
          init: {
            type: "CallExpression",
            callee: { type: "Identifier", name: "hello" },
            arguments: [
              { type: "Literal", value: 34, raw: "34" },
              { type: "Literal", value: null, raw: "world" },
            ],
          },
        },
      ],
    },
  ]);
  const result = interpreter(tree);
  await assertSnapshot(t, result);
});

// varible = null test
