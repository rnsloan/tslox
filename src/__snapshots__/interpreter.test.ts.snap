export const snapshot = {};

snapshot[`Variables 1`] = `
"var hello = 12.45;
var world = null;"
`;

snapshot[`Functions 1`] = `
'foo(true)
var foo = hello(34,"world");'
`;
